### DISCUSSION TITLE
Creating new items from scripts

### DISCUSSION THREAD

**Discussion by DHoelz** [sepinf-inc/IPED#1044] [Q&A]
@lfcnassif,

Could you make a example script (in python or js) showing how to create new itens from scripts?

As I'm not a programmer, I'm having difficulties to implement this functionality. I've managed to add a new file to the case, howewer I don't think I'm doing it the right way and I couldn't set the view file for the newly created item.

**Comment by lfcnassif** [comment]
Hi @DHoelz. Sure. This and next week I'm a bit busy, but I'll try to write an example script for that as soon as possible. What kind of script are you trying to write, a parser or a task?

**Reply by DHoelz** [reply]
Okay, no problem, nothing urgent.

It's a task that executes an external program that parsers the file and creates more than one text report from a single file.

As I need them to be separe reports I couldn't  just use the external parser option IPED has as I believe IPED would set only one view file for that item, right?

**Reply by lfcnassif** [reply]
> I believe IPED would set only one view file for that item, right?

Yes, correct. Maybe that feature could be generalized in the future...

While I don't write the example script, I think possibly you need to add something like this in your code:
https://github.com/sepinf-inc/IPED/blob/3.18.14/iped-engine/src/main/java/dpf/sp/gpinf/indexer/process/task/ParsingTask.java#L627-L635

That will store the item content into the internal case database.

**Reply by lfcnassif** [reply]
> While I don't write the example script, I think possibly you need to add something like this in your code:
> https://github.com/sepinf-inc/IPED/blob/3.18.14/iped-engine/src/main/java/dpf/sp/gpinf/indexer/process/task/ParsingTask.java#L627-L635

Forget above, in 3.18.x you can do something easier. See the example below (this won't work in 4.0):
```
// Create new Item javascript script example.

function getName(){
    return "NewItemExampleTask";
}

function init(confProps, configFolder){
}

function finish(){
}

function process(item){
    
    // skip items to be not processed based on some criteria
    // this will create the new item when processing root
    if (!item.isRoot()) return;
    
    // create a new item and set parent-child relationship
    var Item = Java.type("gpinf.dev.data.Item");
    var newItem = new Item();
    newItem.setParent(item);
    item.setHasChildren(true);

    // set basic properties
    var name = "name.pdf";
    newItem.setName(name);
    newItem.setPath(item.getPath() + "/" + name);
    
    // reference to your external item content
    var File = Java.type("java.io.File");
    var srcFile = new File("F:\\teste-files\\pdf-rotated\\test.pdf");

    // moduleDir is provided as a global var
    var dstPath = "/view/destFile.pdf";
    var dstFile = new File(moduleDir, dstPath);
    
    // copy item content to case folder, so it will be portable
    var Files = Java.type("com.google.common.io.Files");
    dstFile.getParentFile().mkdirs();
    Files.copy(srcFile, dstFile);
    
    // point the new item to the copied content
    var relativePath = moduleDir.getName() + dstPath;
    newItem.setExportedFile(relativePath);
    newItem.setIdInDataSource(relativePath);
    newItem.setLength(dstFile.length());

    // add new item to processing queue, worker is provided as a global var
    worker.processNewItem(newItem);
}

```

Remember to save your script in  %profile%/conf/scripts and install it at %profile%/conf/TaskInstaller.xml

**Reply by lfcnassif** [reply]
PS: You can use any kind of report format with example above (txt, html, pdf, jpg...)

**Reply by lfcnassif** [reply]
I just fixed the script to work properly.

**Reply by DHoelz** [reply]
I'll give it a try and see if I can make it work! 

Thanks!

**Reply by DHoelz** [reply]
Hi @lfcnassif, 

Managed to make the script work creating several new files. 

However I'm having a strange problem. I can't set a new category to the file. I've done that already in other scripts but for some reason it just won't work. I'm using newItem.addCategory('newCategory').

The only way I managed to sort the files in a new category was by changing the file mimetype, using newItem.setMediatypeStr('newMimeType'), and setting the new category on CategoriesByTypeConfig.txt.

Finally, I'm getting an alert from IPED about the new files created that were processed and indexed:

![image](https://user-images.githubusercontent.com/31664860/161549915-39611a94-4dee-4f86-a528-1f4ca2b1f1f1.png)

Is that expected or am I missing something in the script?

**Reply by lfcnassif** [reply]
Hi @DHoelz,

About setting the new category, probably it is being overridden by SetCategoryTask (your new item is sent through processing pipeline to all other tasks), since it uses `setCategory` and not `addCategory` when processing the item. Maybe that could be changed, but I need to think about possible bad effects... The common approach to define a new category is what you did, specifying a new mimeType and configuring it into CategoriesByTypeConfig.txt. Other approach would be to set it in scripts/RefineCategoryTask.js based on the new mimetype.

Regarding the IPED warning about indexed/processed item counts, my fault. You must also call after `caseData.addItem(newItem);` another method: `caseData.incDiscoveredEvidences(1);`. The last isn't called automatically because in some scenarios (when processing large e01/dd images) we don't add all items to caseData/processing queue at once, otherwise processing queue memory usage would be large when processing dozens of millions of files, but we need to count beforehand the total number of items to be processed using `caseData.incDiscoveredEvidences()` to give the user some estimation about processing time to finish.

Instead of calling that two methods `caseData.addItem(newItem);`  and `caseData.incDiscoveredEvidences(1);` a better option would be to call `worker.processNewItem(newItem);` it will call those 2 methods under the hood, I'll update the example script.

**Reply by DHoelz** [reply]
Shouldn't that be prevented by simply putting the script after the SetCategoryTask? 
Anyway the script is working fine and the worker.processNewItem(newItem) fixed the error warning.

Thanks alot for the help!

**Reply by lfcnassif** [reply]
> Shouldn't that be prevented by simply putting the script after the SetCategoryTask?

It should work for existing items, but not for newly created items since they are sent to processing pipeline since the beginning and pass through all enabled tasks. Or maybe you could put another logic in your script to don't skip the newly created items based on the new mimeType and set the category of those new items in your script, in the second pass, when they are passing through the pipeline.

**Reply by DHoelz** [reply]
Ok. I'll take a look on that. Either way the script is working as expected.

**Reply by lfcnassif** [reply]
Great, you are welcome.

**Reply by lfcnassif** [reply]
Hi @DHoelz, I've just written a better example script if you are going to create thousands of subitems: https://github.com/sepinf-inc/IPED/pull/1102#issuecomment-1125314168

The key difference is the call to `extractor.extractFile(is, newItem, item.getLength());` it will store subitem data in internal case database, instead of creating thousands of files into case folder.

**Reply by DHoelz** [reply]
Hi @lfcnassif, 

When you have the time, could you please update the example script for creating new items for using with IPED 4.0.0?

**Reply by lfcnassif** [reply]
sure

**Reply by lfcnassif** [reply]
Hi @DHoelz, now I'm on vacation, if I have time I'll update the example, if not, I can do that when I return in 2 weeks.

**Reply by lfcnassif** [reply]
Until then, you can take a look here:
https://github.com/sepinf-inc/IPED/blob/4.0.0-final/iped-engine/src/main/java/iped/engine/task/video/VideoThumbTask.java#L542

**Reply by DHoelz** [reply]
Hi @lfcnassif, sorry to take so long to reply. 

I did some tests and got a few different results.

First, I have to explain that with IPED 3.18.* I coudn't run more that one python script at the same time. I even tried to run 2 dummy python scripts to see if it was my code but that didn't work either. Seemed to be a problem with JEP.

Because of that I made one big script that runs several tools that creates several new items.

Now, about the changes to make it work with version 4.*:

At first I made the suggested changes on my scripts in version 3.18, using the extractor method.
It worked OK.

After that I tried it with version 4 (with the other necessary changes) but I got some odd results.

**1.** Most of it works, however they keep returning the same error for every new item I create:

2022-08-10 13:20:52	[WARN]	[engine.task.PythonTask]			Exception from EventReport.py on base_testes/jumplist/lenovo/NVMe256GB.E01/GPT_Partition_Table/vol6 [Basic data partition] (237GB)/NTFS/Users/login/AppData/Roaming/Microsoft/Windows/Recent/AutomaticDestinations/d00655d2aa12ff6d.automaticDestinations-ms>>D00655D2AA12FF6D.AUTOMATICDESTINATIONS-MS-Report: jep.JepException: Unable to find object with name: **eventreport_thread_74**
jep.JepException: Unable to find object with name: **eventreport_thread_74**
	at jep.Jep.invoke(Native Method) ~[jep-4.0.3.jar:?]
	at jep.Jep.invoke(Jep.java:257) ~[jep-4.0.3.jar:?]
	at iped.engine.task.PythonTask.process(PythonTask.java:310) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processMonitorTimeout(AbstractTask.java:250) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:180) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.core.Worker.process(Worker.java:177) [iped-engine-4.0.1.jar:?]
	at iped.engine.core.Worker.processNewItem(Worker.java:227) [iped-engine-4.0.1.jar:?]
	at iped.engine.core.Worker.processNewItem(Worker.java:204) [iped-engine-4.0.1.jar:?]
	at jep.Jep.invoke(Native Method) ~[jep-4.0.3.jar:?]
	at jep.Jep.invoke(Jep.java:257) [jep-4.0.3.jar:?]
	at iped.engine.task.PythonTask.process(PythonTask.java:310) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processMonitorTimeout(AbstractTask.java:250) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:180) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.PythonTask.sendToNextTask(PythonTask.java:273) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.sendToNextTask(AbstractTask.java:207) [iped-engine-4.0.1.jar:?]
	at iped.engine.task.AbstractTask.processAndSendToNextTask(AbstractTask.java:188) [iped-engine-4.0.1.jar:?]
	at iped.engine.core.Worker.process(Worker.java:177) [iped-engine-4.0.1.jar:?]
	at iped.engine.core.Worker.run(Worker.java:266) [iped-engine-4.0.1.jar:?]


**2.** One of my scripts seems to get stuck. This is actually just a script that runs regripper several times to make custom reports for different subjects, instead of one huge report for each hive. After that I created a "System Information" category to display the results. The filtering is done in another script file, like the RefineCategoryTask.js

My "workaround" was changing the regripper class in Java as it seemed like a simple change. Btw, if this is interest to you I can try to make a PR.

Btw, just to remember, I'm not a programmer.

**3.** I have another script that creates boorkmarks based on an external text file. This one seems to be an JEP error too:
** I tested the example python script also to see if it was my code, but I got the same error.

jep.JepException: <class 'TypeError'>: **Cannot create a consistent method resolution
order (MRO)** for bases object, Serializable, IBookmarks
	at C:\Users\WERNECK\Desktop\IPED-4.0.1-spi\iped-4.0.0-beta\conf\scripts\PythonScriptTask.finish(PythonScriptTask.py:44)
	at jep.Jep.invoke(Native Method)
	at jep.Jep.invoke(Jep.java:257)
	at iped.engine.task.PythonTask.finish(PythonTask.java:252)
	at iped.engine.core.Worker.finishTasks(Worker.java:135)
	at iped.engine.core.Worker.run(Worker.java:301)


Maybe I'm just missing something.

**Reply by lfcnassif** [reply]
Hi @DHoelz,

> First, I have to explain that with IPED 3.18.* I coudn't run more that one python script at the same time. I even tried to run 2 dummy python scripts to see if it was my code but that didn't work either. Seemed to be a problem with JEP.

Yes, that limitation/bug was fixed in version 4.0 as described in the release notes:
https://github.com/sepinf-inc/IPED/blob/4.0.0-beta/ReleaseNotes.txt#L133

1. Could you share your python script (or other sample script) and triggering data so I could try to reproduce? Actually, I still didn't try to create new items from **python** scripts, I could try, but as I'm working with other stuffs, that would make things easier/faster for me...

2. Hum... very interesting! That looks really useful for users, a PR would be very welcome, thanks! PS: Not sure if I understood correctly, but if you changed the RegRipperParser java class to fix a python issue, that looks pretty weird...

3. Was that thrown with the sample PythonScriptTask.py provided? Have you replaced or installed a different Python or JEP library different than those embedded into IPED-4?

**Reply by lfcnassif** [reply]
Hi @DHoelz, I had some time to do tests here:
1. I wasn't able to reproduce, simple method below works fine:
```
    def newSubItem(self, item):
        from iped.engine.data import Item
        newItem = Item()
        newItem.setParent(item)
        
        newItem.setName('test')
        newItem.setPath(item.getPath() + ">>" + newItem.getName())
        newItem.setSubItem(True)

        # must be incremented for each subitem created from the same parent
        newItem.setSubitemId(0)
        
        from iped.engine.core import Statistics
        Statistics.get().incSubitemsDiscovered();
        
        # this avoid negative time to finish in stats, subitem size shouldn't be sum/added to stats
        newItem.setSumVolume(False);
        
        worker.processNewItem(newItem);
```
If you are creating new threads from python, you shouldn't process new subitems from them, but from the original thread that entered the process(self, item) method. OR, alternatively, replace the last call by `worker.processNewItem(newItem, ProcessTime.Later)` so it will enqueue new items and they will be processed by iped java threads. Not sure if that is the case, I haven't seen your script...

2. I would need to see your script to understand what is the issue and how your workaround worked. Sharing the code in a Draft PR may be enough.

3. This is actually a bug, thanks for reporting. I fixed it on #1274 and also fixed other issues in examples scripts caused by API changes in 4.0, see #1273

**Reply by DHoelz** [reply]
1. I'll try to take a look this afternoon and see what I'm doing wrong. 

2. My "workaround" was actually just modify the java class for regripper, no external scripts involved. I'll try to make the PR but I've never done that before.

3. Thanks! I'll take a look these issues.

**Reply by DHoelz** [reply]
Hi @lfcnassif, after a long time I tried again the python scripts that were giving me the thread exception but I couldn't get them to work correctly 100% of the time. At first they seemed to work with my test base however I realized that they only worked because there were just a few itens, and If I got it right, in these cases you do use the ProcessTime.Later argument.

Using it on a larger set gave me the thread exception error. Tried a lot of things to try to solve it however my knowledge is limited in programming languages. 

Regarding the suggestion of placing the ProcessTime.Later attribute on the worker, apparently I cannot use it because it is an enum variable. At least that's what I understood after some research.

**Reply by lfcnassif** [reply]
Sorry for the delay here.

> there were just a few itens, and If I got it right, in these cases you do use the ProcessTime.Later argument.

**EDITED:** Yes, the idea is: if there are a few items in the queue, enqueue new items so they can be distributed to workers that should become idle soon; if there are many items in the queue, other workers will be busy for some time, so pause processing current (parent) item and process the new subitems in the current worker to take advantage of memory/IO caches (this needs your task to be stateless to avoid overring parent item processing state with its subitems processing state).

> Regarding the suggestion of placing the ProcessTime.Later attribute on the worker, apparently I cannot use it because it is an enum variable. At least that's what I understood after some research.

I'll take a look to check if it is possible to use that java enum from python side.

**Reply by lfcnassif** [reply]
Try this, it worked here now:
```
from iped.engine.core import Worker
worker.processNewItem(newItem, Worker.ProcessTime.LATER);
```

**Reply by DHoelz** [reply]
It worked!
To be honest, I think I tried something similar. Well, it's working now with no errors. 
Thanks!

**Comment by DHoelz** [comment]
@lfcnassif, about issue #1 I made a few tests and that thread error was a result of 3 corrupted files. All other scripts runs fine with no issues at all.

About the regripper custom reports, I commited the changes I made to my repository, however I just realised you import regripper from its repository, and the implementation I made changes the plugins folder directory. I created one subfolder for each hive, and in each subfolder you can put a text file with the plugins you wish the report to have.

Regirpper plugin folder structure:
![image](https://user-images.githubusercontent.com/31664860/184643635-5a5df72c-e5f5-424f-8dbc-eb8e90ea2564.png)


Software folder with custom report files:
![image](https://user-images.githubusercontent.com/31664860/184644034-7c5d368b-5bdd-447b-bf04-bc209765ee90.png)



Should I change the folder location or try to change how the custom reports are made?

**Reply by lfcnassif** [reply]
Hi @DHoelz,
1. I still think this is strange. Corrupted files shouldn't break iped code, unless if they are breaking your script logic or thread handling...
2. Actually we download a custom regripper zip. Maybe we could fork regripper repo, apply your changes in the new fork and use it. Would that work for you?

**Reply by DHoelz** [reply]
@lfcnassif 

1. It seems to be something specific with the external tool I'm using. The script logic is the same for other external tools I'm using and they seem to work fine (atleat till now).

If you have the time I can send you the script and, privately, the jumplist files I'm using. 


2. Yes, that would work. I can make the PR and send you my plugins folder for you to test and see if it's worth the hassle.  Would that be ok?

**Reply by lfcnassif** [reply]
> If you have the time I can send you the script and, privately, the jumplist files I'm using.

Sure.


> 2\. Yes, that would work. I can make the PR and send you my plugins folder for you to test and see if it's worth the hassle.  Would that be ok?

Yes.

**Reply by DHoelz** [reply]
1. Please send me an e-mail or another form of contact I can send you the files.

2. Sent the PR.

**Reply by lfcnassif** [reply]
> 1. Please send me an e-mail or another form of contact I can send you the files.

It's on my profile.

> 2. Sent the PR.

Thank you!
